### ISSUE TITLE
Macro string literal completion edits not replacing prior text

### ISSUE THREAD

**Issue by RoloEdits** [rust-lang/rust-analyzer#17479]
**rust-analyzer version**: rust-analyzer 0.0.0 (498b94b730 2024-06-17)

**rustc version**: rustc 1.74.0 (79e9716c9 2023-11-13)

**editor or extension**: helix

**code snippet to reproduce**: 
```rust
// add your code here
use chrono::serde::ts_milliseconds;
use chrono::{DateTime, Utc};
use serde::Deserialize;

#[derive(Deserialize)]
struct A {
    #[serde(with = "ts_mill")]
    time: DateTime<Utc>,
}

fn main() {
    println!("Hello, world!");
}
```
![helix_auto_complete_in_string](https://github.com/rust-lang/rust-analyzer/assets/12489689/617dc389-7ef7-47e5-93d5-71d2937149ca)

Originally thought it was a helix issue, but was directed here after some discussion.
https://github.com/helix-editor/helix/discussions/11020

**Comment by Veykril** [comment]
Hmm, we shouldn't trigger those there at all actually given we are inside a string token. It's not really possibly for us to treat this completion correctly inside a string.

**Comment by BLANKatGITHUB** [comment]
I would like to try fixing this, @Veykril l can you please help a little with relevant code and some instruction if possible as it would be my first contribution. I would appreciate it.

**Comment by ChayimFriedman2** [comment]
@BLANKatGITHUB If you want to disable completion inside strings, this should be simple - check if `ctx.original_token` is a string [here](https://github.com/rust-lang/rust-analyzer/blob/ba1c9146fcca65fa9c40d1e4e73bdcb898a8de46/crates/ide-completion/src/lib.rs#L178).

**Comment by Veykril** [comment]
I think we need to check if `original_token` and `self_token` (bad name, this is the token in the expansion) here https://github.com/rust-lang/rust-analyzer/blob/ba1c9146fcca65fa9c40d1e4e73bdcb898a8de46/crates/ide-completion/src/context/analysis.rs#L278-L283
differ in terms of what they are.

So a basic check on equality of their syntax kind ought to be enough + if that check fails (that is they differ) do another check if both of them are identifiers (as keywords get their own kind but we want to treat that as identifiers), ie. in pseudo-ish code if `original_token.kind() != self_token.kind() && is_ident_like(original_token.kind()) != is_ident_like(self_token.kind())`, we should break out and dont do completions, we have an is_ident_like function somehwere (might be a method on `SyntaxKind`

**Comment by BLANKatGITHUB** [comment]
> I think we need to check if `original_token` and `self_token` (bad name, this is the token in the expansion) here
> 
> [rust-analyzer/crates/ide-completion/src/context/analysis.rs](https://github.com/rust-lang/rust-analyzer/blob/ba1c9146fcca65fa9c40d1e4e73bdcb898a8de46/crates/ide-completion/src/context/analysis.rs#L278-L283)
> 
> Lines 278 to 283 in [ba1c914](/rust-lang/rust-analyzer/commit/ba1c9146fcca65fa9c40d1e4e73bdcb898a8de46)
> 
>  fn analyze( 
>      sema: &Semantics<'_, RootDatabase>, 
>      expansion_result: ExpansionResult, 
>      original_token: &SyntaxToken, 
>      self_token: &SyntaxToken, 
>  ) -> Option<(CompletionAnalysis, (Option<Type>, Option<ast::NameOrNameRef>), QualifierCtx)> { 
> 
> differ in terms of what they are.
> So a basic check on equality of their syntax kind ought to be enough + if that check fails (that is they differ) do another check if both of them are identifiers (as keywords get their own kind but we want to treat that as identifiers), ie. in pseudo-ish code if `original_token.kind() != self_token.kind() && is_ident_like(original_token.kind()) != is_ident_like(self_token.kind())`, we should break out and dont do completions, we have an is_ident_like function somehwere (might be a method on `SyntaxKind`

I tried that but it failed 2 tests
'''
failures:
    render::macro_::tests::complete_missing_macro_arg
    tests::expression::in_macro_expr_frag

test result: FAILED. 573 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 10.91s
'''

**Comment by BLANKatGITHUB** [comment]
> [@BLANKatGITHUB](https://github.com/BLANKatGITHUB) If you want to disable completion inside strings, this should be simple - check if `ctx.original_token` is a string [here](https://github.com/rust-lang/rust-analyzer/blob/ba1c9146fcca65fa9c40d1e4e73bdcb898a8de46/crates/ide-completion/src/lib.rs#L178).

Thanks for help but its still kind of failing 2 test's, I would appreciate your help as its my first time contributing.

**Comment by ChayimFriedman2** [comment]
What do the failed tests print?

**Comment by BLANKatGITHUB** [comment]
What do the failed tests print?

I kind of implemented

`if ctx.original_token == syntax::Syntax::STRING{
return None
}`

it failed 2 tests now I am trying kind of simple code
`    if trigger_character == Some('"') {
        return None;
    }`
Sorry if I am doing anything wrong and taking your time,
I have never worked on big project like this so I dont understand much of code base, but I would like to help the project if I can , I hope you can guide me

**Comment by BLANKatGITHUB** [comment]
> I think we need to check if `original_token` and `self_token` (bad name, this is the token in the expansion) here
> 
> [rust-analyzer/crates/ide-completion/src/context/analysis.rs](https://github.com/rust-lang/rust-analyzer/blob/ba1c9146fcca65fa9c40d1e4e73bdcb898a8de46/crates/ide-completion/src/context/analysis.rs#L278-L283)
> 
> Lines 278 to 283 in [ba1c914](/rust-lang/rust-analyzer/commit/ba1c9146fcca65fa9c40d1e4e73bdcb898a8de46)
> 
>  fn analyze( 
>      sema: &Semantics<'_, RootDatabase>, 
>      expansion_result: ExpansionResult, 
>      original_token: &SyntaxToken, 
>      self_token: &SyntaxToken, 
>  ) -> Option<(CompletionAnalysis, (Option<Type>, Option<ast::NameOrNameRef>), QualifierCtx)> { 
> 
> differ in terms of what they are.
> So a basic check on equality of their syntax kind ought to be enough + if that check fails (that is they differ) do another check if both of them are identifiers (as keywords get their own kind but we want to treat that as identifiers), ie. in pseudo-ish code if `original_token.kind() != self_token.kind() && is_ident_like(original_token.kind()) != is_ident_like(self_token.kind())`, we should break out and dont do completions, we have an is_ident_like function somehwere (might be a method on `SyntaxKind`

I cant find 'is_ident_like'  can you please help a little more , I am trying both suggestion given to me in this issue, 
I kind of implemented one and submitted for review I am not sure about it so help fom expierenced people will be appreciated

**Comment by Veykril** [comment]
Its this one https://github.com/rust-lang/rust-analyzer/blob/597d8e837a0a1a48349164d4057f6e42765d7a32/crates/parser/src/syntax_kind.rs#L34

**Comment by BLANKatGITHUB** [comment]
Umm I added 
`if (original_token.kind() != self_token.kind()) 
    && (SyntaxKind::is_any_identifier(original_token.kind()) != SyntaxKind::is_any_identifier(self_token.kind())) {
    return None;
}`

but it failed some tests


'failures:
    render::macro_::tests::complete_missing_macro_arg
    tests::expression::in_macro_expr_frag '

In first step it was unwrapping a None value,
and in second there was no completion for non string items

**Comment by vishruth-thimmaiah** [comment]
> So a basic check on equality of their syntax kind ought to be enough + if that check fails (that is they differ) do another check if both of them are identifiers (as keywords get their own kind but we want to treat that as identifiers), ie. in pseudo-ish code if `original_token.kind() != self_token.kind() && is_ident_like(original_token.kind()) != is_ident_like(self_token.kind())`, we should break out and dont do completions, we have an is_ident_like function somehwere (might be a method on `SyntaxKind`

@Veykril - I attempted to implement your suggested solution, but it wasn't successful. Instead, I found a working alternative approach: comparing the token types by checking if original_token.kind() is a string while self_token.kind() is not. Could you please review if this is a valid approach or if there might be edge cases I should consider?

**Comment by vishruth-thimmaiah** [comment]
> I attempted to implement your suggested solution, but it wasn't successful. 

I did some digging, and looks like commit - 15d2d509d0fff1e46c5a7db16e7a692ee26ae029 is the reason your suggestion isn't working - for the code snippet provided, the value of self_token is COLON2, instead of IDENT.

**Comment by Veykril** [comment]
Oh my condition is wrong, that should be `!(original_token.kind() == self_token.kind() || (is_ident_like(original_token.kind()) && is_ident_like(self_token.kind())))` or with the negation pulled further in `original_token.kind() != self_token.kind() && !(is_ident_like(original_token.kind()) && is_ident_like(self_token.kind()))`
